src/chrome/utils/localizedError.ts

```diff
@@ -0,0 +1,119 @@
+}
+
+/** Format of the bundle.json file */
+interface LocalizationBundle {
+    [file: string]: { keys: (string | undefined)[] } | undefined;
+}
+
+/** We store the bundle metadata on memory to be able to extract the localization keys based on the file and message index */
+let bundle: LocalizationBundle = {};
+const bundlePath = path.resolve(__dirname, '../../../nls.metadata.json');
+const bundleFolderPath = path.dirname(bundlePath);
+try {
+    bundle = JSON.parse(readFileSync(bundlePath, 'utf8'));
+} catch (exception) {
+    telemetry.reportError('Failed to read bundle', exception);
+}
+
+/** Given the localization key, and the file, do our best effort to return the localization key (We get it from the bundle).
+ * In dev we'll normally get the localization key as a string
+ * In prod, the nls library replaces the key with a number which is the index in the bundle that has that particular message, so we need to use the file,
+ * the index, and the bundle to find the localization key
+ */
+function extractLocalizationKey(key: string | number | nls.LocalizeInfo, file: string | undefined): string {
+    if (typeof key === 'number') {
+        if (file !== undefined) {
+            const fileNormalizedPath = normalizeFilePath(file);
+            const messagesForFile = bundle[fileNormalizedPath];
+            if (messagesForFile !== undefined) {
+                const textKey = messagesForFile.keys[key];
+                if (textKey !== undefined) {
+                    return textKey;
+                }
+            }
+        }
+
+        return `${file}:${key}`;
+    } else if (typeof key === 'string') {
+        return key;
+    } else {
+        return 'getNextLastKey parameter was unexpected';
+    }
+}
+
+/** Normalize the file path to the format used in the localization bundles (Relative path, no extension, and forward slashes) */
+function normalizeFilePath(file: string): string {
+    const fileRelativePath = path.relative(bundleFolderPath, file);
+    const normalizedPath = fileRelativePath.replace(/\\/g, '/').replace(/\.js$/g, '');
+    return normalizedPath;
+}
+
+const originalConfig = nls.config;
+let currentLocale: string | undefined = undefined;
+
+/** We replace nls.config with our own version so we can know which locale is set, and force all the localize functions to be re-created when needed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).config = (opts?: nls.Options) => {
+    currentLocale = _.defaultTo(opts, { locale: undefined }).locale;
+    return originalConfig(opts);
+};
+
+const originalLoadMessageBundle = nls.loadMessageBundle;
+/** We replace nls.loadMessageBundle with our own custom version, which will re-created the localize function after the locale has been changed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).loadMessageBundle = (file?: string) => {
```

Comment:
I don't understand this either

Reply:
From what I've experimented of nls, when we call 
nls.loadMessageBundle() on a file the first time, it returns a localize function that works in English.

After we call nls.config on the initialize message, all the localize functions are already created, and tied to the English language, so the translations are not being used. To get the translations to be used, we need to call nls.loadMessageBundle() for all files again, to get a new version of localize that is tied to the proper locale.

This methods overrides nls.loadMessageBundle() to automatically re-create the localize function after a locale is specified in nls.config

Reply:
Like this works in v1. So why are weird hacks needed? If there is a bug in the library let's fix that.